go/vt/vtqueryserver/endtoend_test.go

```diff
@@ -0,0 +1,314 @@
+	t.Helper()
+
+	numQueries := tabletenv.MySQLStats.Count()
+	result, err := conn.ExecuteFetch(sql, 1000, false)
+	if err != nil {
+		t.Errorf("error: %v", err)
+	}
+	numQueries = tabletenv.MySQLStats.Count() - numQueries
+
+	if numQueries != expectedNumQueries {
+		t.Errorf("expected %d mysql queries but got %d", expectedNumQueries, numQueries)
+	}
+
+	if result.RowsAffected != expectedRowsAffected {
+		t.Errorf("expected %d rows affected but got %d", expectedRowsAffected, result.RowsAffected)
+	}
+}
+
+func TestQueries(t *testing.T) {
+	ctx := context.Background()
+	conn, err := mysql.Connect(ctx, &proxyConnParams)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Try a simple query case.
+	testFetch(t, conn, "select * from test", 0)
+
+	// Try a simple error case.
+	_, err = conn.ExecuteFetch("select * from aa", 1000, true)
+	if err == nil || !strings.Contains(err.Error(), "table aa not found in schema") {
+		t.Fatalf("expected error but got: %v", err)
+	}
+}
+
+func TestAutocommitDMLs(t *testing.T) {
+	ctx := context.Background()
+
+	conn, err := mysql.Connect(ctx, &proxyConnParams)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	conn2, err := mysql.Connect(ctx, &proxyConnParams)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	testDML(t, conn, "insert into test (id, val) values(1, 'hello')", 3, 1)
+
+	testFetch(t, conn, "select * from test", 1)
+	testFetch(t, conn2, "select * from test", 1)
+
+	testDML(t, conn, "delete from test", 4, 1)
+
+	testFetch(t, conn, "select * from test", 0)
+	testFetch(t, conn2, "select * from test", 0)
+}
+
+func TestPassthroughDMLs(t *testing.T) {
+	ctx := context.Background()
+
+	queryServer.SetPassthroughDMLs(true)
+	conn, err := mysql.Connect(ctx, &proxyConnParams)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	conn2, err := mysql.Connect(ctx, &proxyConnParams)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	testDML(t, conn, "insert into test (id, val) values(1, 'hello')", 3, 1)
+	testDML(t, conn, "insert into test (id, val) values(2, 'hello')", 3, 1)
+	testDML(t, conn, "insert into test (id, val) values(3, 'hello')", 3, 1)
+
+	// Subquery DMLs are errors in passthrough mode with SBR, unless
+	// SetAllowUnsafeDMLs is set
+	_, err = conn.ExecuteFetch("update test set val='goodbye'", 1000, true)
+	if err == nil || !strings.Contains(err.Error(), "cannot identify primary key of statement") {
+		t.Fatalf("expected error but got: %v", err)
+	}
+
+	queryServer.SetAllowUnsafeDMLs(true)
+
+	// This is 3 queries in passthrough mode and not 4 queries as it would
+	// be in non-passthrough mode
+	testDML(t, conn, "update test set val='goodbye'", 3, 3)
+
+	testFetch(t, conn, "select * from test where val='goodbye'", 3)
+	testFetch(t, conn2, "select * from test where val='goodbye'", 3)
+
+	testDML(t, conn, "delete from test", 4, 3)
+
+	testFetch(t, conn, "select * from test", 0)
+	testFetch(t, conn2, "select * from test", 0)
+}
+
+func TestTransactions(t *testing.T) {
```

Comment:
I couldn't see a test where a begin->dmls->begin causes the previous transaction to be commited and a new one started.